includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php

```diff
@@ -0,0 +1,234 @@
+
+/**
+ * Class Newspack_Newsletters_Letterhead
+ */
+final class Newspack_Newsletters_Letterhead extends \Newspack_Newsletters_Service_Provider {
+	const LETTERHEAD_WP_OPTION_KEY = 'newspack_newsletters_letterhead_api_key';
+
+	/**
+	 * Gets the Letterhead API credentials. Right now, that's just a key : ).
+	 *
+	 * @return string The user's LH API key.
+	 */
+	public function api_credentials() {
+		$letterhead_key = get_option( self::LETTERHEAD_WP_OPTION_KEY, '' );
+		return $letterhead_key;
+	}
+
+	/**
+	 * Fetch the newsletter's promotions from Letterhead and format them for insertion
+	 * in Newspack's email template.
+	 *
+	 * @param string $date The date of the publication.
+	 * @param int    $length_of_newsletter The length of the newsletter.
+	 * @return array
+	 */
+	public function get_and_prepare_promotions_for_insertion( $date, $length_of_newsletter ) {
+		$promotions_from_api = $this->get_promotions_by_date( $date );
+
+		return array_map(
+			function ( Newspack_Newsletters_Letterhead_Promotion $promotion ) use ( $length_of_newsletter ) {
+				return $promotion->convert_to_compatible_newspack_ad_array( $length_of_newsletter );
+			},
+			$promotions_from_api
+		);
+	}
+
+	/**
+	 * This will call the Letterhead promotions API with the specific date passed as the
+	 * argument and the appropriate credentials. It will return an array.
+	 *
+	 * @param string $date The date of publication.
+	 * @return array
+	 */
+	public function get_promotions_by_date( $date ) {
+		if ( ! $this->has_api_credentials() ) {
+			return [];
+		}
+
+
+		$credentials = $this->api_credentials();
+		$url         = NEWSPACK_NEWSLETTERS_LETTERHEAD_ENDPOINT . "/api/v2/promotions/?date={$date}&mjml=true";
+
+		$request_headers = [
+			'Authorization' => "Bearer {$credentials}",
+		];
+
+		$request_arguments = [
+			'headers' => $request_headers,
+		];
+
+		$response = function_exists( 'vip_safe_wp_remote_get' )
+			? vip_safe_wp_remote_get( $url, '', 1, 1, 60, $request_arguments )
+			// phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.wp_remote_get_wp_remote_get
+			: wp_remote_get( $url, $request_arguments );
+
+		/**
+		 * Our end users don't really benefit from having a broken experience just because our API call
+		 * failed. We can fail silently and return an empty array.
+		 */
+		if ( is_wp_error( $response ) ) {
+			// We should log the error somewhere.
+			return [];
+		}
+
+		$promotions = wp_remote_retrieve_body( $response );
+
+		return $this->get_promotions_from_json_response( $promotions );
+	}
+
+	/**
+	 * Pass the json response from, presumably, a Letterhead API call, and return an array of
+	 * Promotions.
+	 *
+	 * @param string $promotions_json The json response from a successful Lettherhead API query.
+	 * @return array Newspack_Newsletters_Letterhead_Promotion[]
+	 */
+	private function get_promotions_from_json_response( $promotions_json ) {
+		$promotion_response_object = json_decode( $promotions_json, false );
```

Comment:
In my testing, I got `Nada.` string as a response (this happened right after opening the editor with a new post). I suppose an empty array was intended?

Reply:
![image](https://user-images.githubusercontent.com/1489144/114483141-db07e880-9bd5-11eb-9a91-6b91583889db.png)

When I make the next pull maybe we can coordinate and you can show me if it fails again : ). After merging master down and making some of the changes you suggested, I was able to load a promotion - it might have been that the account you were pointing to just didn't have a promotion set for that date, but there's a lot of reasons it could turn up goose-eggs : ). If there aren't any promotions on that day it will return an empty array, yup.

Reply:
> If there aren't any promotions on that day it will return an empty array, yup.

Hm, I'm still getting a `Nada.` response with no upcoming promotions :/

Reply:
I'll touch base with ya and make sure we're hitting the right endpoint.